src/agentscope/agents/web_voyager_agent.py

```diff
@@ -0,0 +1,411 @@
+                or hard-coded in the agent.tion.
+            memory_config (`Optional[dict]`):
+                The config of memory.
+        """
+        super().__init__(
+            name=name,
+            sys_prompt=sys_prompt,
+            model_config_name=model_config_name,
+            use_memory=True,
+            memory_config=memory_config,
+        )
+        self.browser = browser
+        self.max_iter = max_iter
+        self.max_attached_imgs = max_attached_imgs
+        self.default_homepage = default_homepage
+        self.task_question = ""
+        self.task_dir = ""
+        self.task_web = ""
+        self.task_answer = ""
+
+    def _clip_message_and_obs(self, msg_list: list) -> list:
+        """
+        Avoid to much image contents so the model won't confuse.
+        """
+        max_img_num = self.max_attached_imgs
+        clipped_msg = []
+        img_num = 0
+        for idx in range(len(msg_list)):
+            curr_msg = msg_list[len(msg_list) - 1 - idx]
+            if curr_msg.role != "user":
+                clipped_msg = [curr_msg] + clipped_msg
+            else:
+                if curr_msg.url is None:
+                    clipped_msg = [curr_msg] + clipped_msg
+                elif img_num < max_img_num:
+                    img_num += 1
+                    clipped_msg = [curr_msg] + clipped_msg
+                else:
+                    msg_omit = (
+                        curr_msg.content.split("Observation:")[0].strip()
+                        + "Observation: A screenshot and some texts. (Omitted in context.)"  # noqa
+                    )
+                    curr_msg_clip = Msg(
+                        name=curr_msg.name,
+                        content=msg_omit,
+                        role=curr_msg.role,
+                    )
+                    clipped_msg = [curr_msg_clip] + clipped_msg
+        return clipped_msg
+
+    def _get_step_msg(
+        self,
+        it: int,
+        warn_obs: str,
+        image_url: str,
+        web_text: str,
+    ) -> Msg:
+        """
+        Get the step message for the current iter.
+        """
+        if it == 1:
+            init_msg = INIT_MSG_PROMPT.format(
+                task_question=self.task_question,
+                task_web_url=self.task_web,
+                web_text=web_text,
+            )
+            init_msg_format = Msg(
+                name="user",
+                content=init_msg,
+                role="user",
+                url=image_url,
+            )
+            return init_msg_format
+        else:
+            obs_text = OBS_MSG_PROMPT.format(
+                warn_obs=warn_obs,
+                web_text=web_text,
+            )
+            curr_msg = Msg(
+                name="user",
+                content=obs_text,
+                role="user",
+                url=image_url,
+            )
+            return curr_msg
+
+    def reply(self, x: dict = None) -> dict:
```

Comment:
to align the type with the latest version `Optional[Union[Msg, Sequence[Msg]]]`

Reply:
Changed